File: JavaRepo/lab2/src/model/CloseRFile.java
package model;

import exceptions.DivideByZeroException;
import exceptions.InterpreterException;
import exceptions.InvalidOperandException;
import exceptions.NotExistingException;
import model.Expressions.Exp;
import model.Statement.IStmt;
import utils.FileData;
import utils.adt.MyIDictionary;
import utils.adt.MyIFileTable;
import utils.PrgState;
import utils.adt.MyIHeap;

import java.io.BufferedReader;
import java.io.IOException;

public class CloseRFile implements IStmt {
    private Exp expFileId;

    public CloseRFile(Exp _expFileId){
        expFileId = _expFileId;
    }

    @Override
    public PrgState execute(PrgState state) throws DivideByZeroException, InvalidOperandException, NotExistingException {
        MyIDictionary<String, Integer> dict = state.getSymTable();
        MyIFileTable<Integer, FileData> fileTable = state.getFileTable();
        MyIHeap<Integer, Integer> heap = state.getHeap();
        int id = expFileId.eval(dict, heap);
        FileData fd = fileTable.get(id);

        try{
            BufferedReader br = fd.getHeader();
            br.close();
            fileTable.remove(id);
        }
        catch (IOException x)
        {
            throw new InterpreterException(x.toString());
        }

        return state;
    }

    @Override
    public String toString(){
        return "CloseFile(" + expFileId + ")";
    }
}

File: Year 2/Semester 1/MAP/C# - Toy language/Lab10/Toy Language/Toy Language/model/statement/CloseRFile.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Toy_Language.model.expressions;
using Toy_Language.utils;
using Toy_Language.utils.adt;

namespace Toy_Language.model.statement
{
    public class CloseRFile : IStmt
    {
        private Exp expFileId;

        public CloseRFile(Exp _expFileId)
        {
            expFileId = _expFileId;
        }

        public PrgState Execute(PrgState state)
        {
            IModDictionary<string, int> symbTable = state.SymbTable;
            int id = expFileId.Eval(symbTable);

            FileData fd = state.FileTable.Get(id);
            System.IO.StreamReader reader = fd.Header;
            reader.Close();
            state.FileTable.Remove(id);

            return state;
        }

        public override string ToString()
        {
            return "CloseFile(" + expFileId + ")";
        }
    }
}
